### DISCUSSION TITLE
pygame black screen

### DISCUSSION THREAD

**Discussion by nkmin7** [community/community#58287] [Programming Help]
### Body

I'm trying to implement Newton's cannon thought experiment in pygame.
Pygame is already installed. But when I run this code, I get a black screen in pygame.. Help me!

```
import pygame
import sys

WIDTH = 800
HEIGHT = 600
RED = (255, 0, 0)
BLACK = (0, 0, 0)
GRAVITY = 9.8

pygame.init()
screen = pygame.display.set_mode((WIDTH, HEIGHT))
clock = pygame.time.Clock()

class Cannon(pygame.sprite.Sprite):
    def __init__(self):
        pygame.sprite.Sprite.__init__(self)
        self.image = pygame.Surface((40, 20))
        self.image.fill(BLACK)
        self.rect = self.image.get_rect()
        self.rect.center = (WIDTH // 2, HEIGHT - 30)
    def update(self):
        mouse_pos = pygame.mouse.get_pos()
        self.rect.centerx = mouse_pos[0]

class Ball(pygame.sprite.Sprite):
    def __init__(self, cannon):
        pygame.sprite.Sprite.__init__(self)
        self.image = pygame.Surface((20, 20))
        self.image.fill(RED)
        self.rect = self.image.get_rect()
        self.rect.center = cannon.rect.center
        self.velocity = 0
    def update(self):
        self.rect.y -= self.velocity
        self.velocity += GRAVITY
        if self.rect.y > HEIGHT:
            self.rect.center = cannon.rect.center
            self.velocity = 0

cannon = Cannon()
ball = None
all_sprites = pygame.sprite.Group()
all_sprites.add(cannon)

while True:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit()
            sys.exit()
        elif event.type == pygame.KEYDOWN:
            if event.key == pygame.K_SPACE:
                if ball is None:
                    ball = Ball(cannon)
                    all_sprites.add(ball)
    all_sprites.update()
    screen.fill((255, 255, 255))
    all_sprites.draw(screen)
    pygame.display.flip()
    clock.tick(60)
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by the-officialjosh** [comment]
I could not really find an error in your code you could try, however - To check for error messages: Run the code in a terminal or IDE that displays error messages. Look for any error messages or exceptions that might provide insights into the cause of the black screen.

Verify Pygame installation: Ensure that you have the correct version of Pygame installed and that it is compatible with your Python environment.

**Comment by nbird33** [comment]
You need to wrap the code inside a while loop that will continue until the game is exited, not just while True. This should get things to display, let me know if it works just had a similar thing happen to me.


running = True
while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
        elif event.type == pygame.KEYDOWN:
            if event.key == pygame.K_SPACE:
                if ball is None:
                    ball = Ball(cannon)
                    all_sprites.add(ball)

**Comment by advayc** [comment]
most likely reason for the black screen is that the display is not being updated correctly within the game loop

try this code

import pygame
import sys

WIDTH = 800
HEIGHT = 600
RED = (255, 0, 0)
BLACK = (0, 0, 0)
GRAVITY = 0.5  # Adjusted gravity for smoother motion

pygame.init()
screen = pygame.display.set_mode((WIDTH, HEIGHT))
clock = pygame.time.Clock()

class Cannon(pygame.sprite.Sprite):
    def __init__(self):
        pygame.sprite.Sprite.__init__(self)
        self.image = pygame.Surface((40, 20))
        self.image.fill(BLACK)
        self.rect = self.image.get_rect()
        self.rect.center = (WIDTH // 2, HEIGHT - 30)
    def update(self):
        mouse_pos = pygame.mouse.get_pos()
        self.rect.centerx = mouse_pos[0]

class Ball(pygame.sprite.Sprite):
    def __init__(self, cannon):
        pygame.sprite.Sprite.__init__(self)
        self.image = pygame.Surface((20, 20))
        self.image.fill(RED)
        self.rect = self.image.get_rect()
        self.rect.center = cannon.rect.center
        self.velocity = 0
    def update(self):
        self.rect.y -= self.velocity
        self.velocity += GRAVITY
        if self.rect.y > HEIGHT:
            self.rect.center = cannon.rect.center
            self.velocity = 0

cannon = Cannon()
ball = None
all_sprites = pygame.sprite.Group()
all_sprites.add(cannon)

while True:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit()
            sys.exit()
        elif event.type == pygame.KEYDOWN:
            if event.key == pygame.K_SPACE:
                if ball is None:
                    ball = Ball(cannon)
                    all_sprites.add(ball)

    # Update objects and screen
    all_sprites.update()
    screen.fill((255, 255, 255))
    all_sprites.draw(screen)
    pygame.display.flip()
    clock.tick(60)
